csharp/src/Drivers/Databricks/CloudFetch/CloudFetchDownloadManager.cs

```diff
@@ -0,0 +1,310 @@
+            }
+
+            try
+            {
+                return await _downloader.GetNextDownloadedFileAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                // If there's an error in the downloader, check if there's also an error in the fetcher
+                if (_resultFetcher.HasError)
+                {
+                    throw new AggregateException("Errors in download pipeline", new[] { ex, _resultFetcher.Error! });
+                }
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task StartAsync()
+        {
+            ThrowIfDisposed();
+
+            if (_isStarted)
+            {
+                throw new InvalidOperationException("Download manager is already started.");
+            }
+
+            // Create a new cancellation token source
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // Start the result fetcher
+            await _resultFetcher.StartAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
+
+            // Start the downloader
+            await _downloader.StartAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
+
+            _isStarted = true;
+        }
+
+        /// <inheritdoc />
+        public async Task StopAsync()
+        {
+            if (!_isStarted)
+            {
+                return;
+            }
+
+            // Cancel the token to signal all operations to stop
+            _cancellationTokenSource?.Cancel();
+
+            // Stop the downloader
+            await _downloader.StopAsync().ConfigureAwait(false);
+
+            // Stop the result fetcher
+            await _resultFetcher.StopAsync().ConfigureAwait(false);
+
+            // Dispose the cancellation token source
+            _cancellationTokenSource?.Dispose();
+            _cancellationTokenSource = null;
+
+            _isStarted = false;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            // Stop the pipeline
+            StopAsync().GetAwaiter().GetResult();
+
+            // Dispose the HTTP client
+            _httpClient.Dispose();
+
+            // Dispose the cancellation token source if it hasn't been disposed yet
+            _cancellationTokenSource?.Dispose();
+            _cancellationTokenSource = null;
+
+            // Mark the queues as completed to release any waiting threads
+            _downloadQueue.CompleteAdding();
+            _resultQueue.CompleteAdding();
+
+            // Dispose any remaining results
+            foreach (var result in _resultQueue.GetConsumingEnumerable(CancellationToken.None))
+            {
+                if (result != EndOfResultsGuard.Instance)
+                {
+                    result.Dispose();
+                }
```

Comment:
Given that `EndOfResultsGuard.Dispose()` is a nop, we could just `.Dispose()` without the `if` guard.